broadinstitute/single_cell_portal_core
# Upgrading to rails-baseimage:2.3.0, Ruby 3.2, Nodejs 20, Vite 4

#### BACKGROUND & CHANGES
This updates to the latest release of `singlecellportal/rails-baseimage`, which updates to Ruby 3.2.2 and Nodejs 20.  Additionally, this updates to Vite 4.5.  It also removes a deprecated gem (`ruby-prof-flamegraph`) that no longer builds past Ruby 3.1.

EDIT: CI will not run on this PR until the [associated upstream PR](https://github.com/broadinstitute/scp-rails-baseimage/pull/29) to update the base image merges.

#### MANUAL TESTING
Given the headaches I went through to manually validate that this works in Docker, I highly recommend that you _not_ do this, especially since we do not develop in Docker locally as part of normal workflows.  There is a bug in Docker Desktop for Mac prior to v4.30 that results in segmentation faults for Vite, and additionally there's a lot of manual shenanigans required to include the required Docker images as they haven't been published yet.  Also, it is not possible to deploy different Docker image tags to staging, so this also cannot be done there.  The underlying scripts accept the arguments, but the Github Action workflows do not allow you to specify them.

Instead, I would validate that the normal non-Dockerized setup still functions:

1. Install Nodejs 20 via nvm:
```
nvm install 20.13.0
...
$ node -v                                                                                       
v20.13.0
```
2. Install Ruby 3.2.2 locally using your manager of choice (this will take a while!)
EDIT: If you have an Apple Silicon Mac and encounter issues installing Ruby 3.2.2, see this thread: https://github.com/rbenv/ruby-build/discussions/2293
Specifically, `MAKE_OPTS='-j2' rbenv install 3.2.2` may solve compiler issues preventing installation.  You may also have to re-install Xcode and the command line tools after requesting an administrator session:
```
sudo rm -rf /Library/Developer/CommandLineTools
sudo xcode-select --install
# wait for a while, click stuff
sudo xcode-select -s /Library/Developer/CommandLineTools
```
Then to install Ruby 3.2.2:
```
# rvm
rvm get stable
rvm install 3.2.2

# rbenv
brew update && brew upgrade ruby-build 
rbenv install 3.2.2
```
3. Make sure that you have the correct version of Ruby loaded in your shell:
```
# rvm
rvm use 3.2.2

# rbenv
rbenv shell 3.2.2

...
$ ruby -v
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [x86_64-darwin23]
```
4. Pull this branch and install necessary gems/packages
```
bundle install
yarn install
```
5. Boot the server & vite as normal and confirm everything loads (make sure the vite shell has the right node/ruby loaded as well).
